framework/test/bundles/libH/TestBundleH.cpp

```diff
@@ -187,6 +168,50 @@ class TestBundleHServiceFactoryReturnsNullPtr : public ServiceFactory
 
 };
 
+// Simulate an error condition whereby the service factory recursively tries to get the same service for the same bundle
+class TestBundleHServiceFactoryRecursion : public ServiceFactory
+{
+public:
+
+  InterfaceMapConstPtr GetService(const Bundle& caller, const ServiceRegistrationBase& sReg)
+  {
+    static bool validRecursion = false;
```

Comment:
Thanks for the explanation @saschazelzer. I think I understand the difference now.

The returned `nullptr` I was referring to is the value returned when the framework breaks out of the infinite recursion, not the `return nullptr` statement in this function. (See line#204 of ServiceFactoryTest.cpp).

More comments would certainly help. Also, It would be cleaner if we could get rid of the unreachable statement `return nullptr`. The loop  and the unreachable return statement can be avoided with the following block.
```
if (caller.GetBundleId() > 0)
{
  auto systemBundle = caller.GetBundleContext().GetBundle(0);
  return systemBundle.GetBundleContext().GetService(sReg.GetReference());
}
```